### PR TITLE
feat: experimental hatchling builder

### PR DESCRIPTION
**Pull request by henryiii**
In progress.

### CODE REVIEW

**Comment by henryiii** [comment]
@aryamanjeendgar, it's passing now, so ready for you to work on it.

File: tests/constraints.txt [hunk 1]
```diff
@@ -1,12 +1,13 @@
 cattrs==22.2.0
+hatchling==0.25.0
```
Comment (ofek) [inline tests/constraints.txt:2]: Is this to support Python 2? If so, note that there is a patch release after this one.

Comment (henryiii) [reply tests/constraints.txt:2]: No, 3.7 is our oldest. Was just trying to pick the oldest one easy to support as a minimum. Can change if you'd like.

**Comment by henryiii** [comment]
I've pulled the API change out so we can ship that for third party plugins sooner. #697

**Comment by ofek** [comment]
I know you're not finished but don't forget that I implemented the [dynamic dependencies](https://hatch.pypa.io/dev/plugins/build-hook/reference/#hatchling.builders.hooks.plugin.interface.BuildHookInterface.dependencies) feature you requested!

**Comment by henryiii** [comment]
TODO: self.directory is `./dist`, not a temp build directory. We'll have to make one ourselves, and clean it in finalize & if there's an error.

File: docs/plugins/hatchling.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+:::
+
+## Basic usage
+
+To use the plugin, make sure hatchling and scikit-build-core are in your
+`build-system.requires`. A recent version of hatchling is best; you need 1.23 to
+get `cmake` and `ninja` auto-added if needed, and 1.24 if you want to write out
+custom scripts, metadata, or shared data.
+
+You need a `tool.hatch.build.targrst.wheel.hooks.scikit-build` section to
```
Comment (henryiii) [inline docs/plugins/hatchling.md:21]:
```suggestion
You need a `tool.hatch.build.targets.wheel.hooks.scikit-build` section to
```

File: docs/plugins/hatchling.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+- You need to leave `cmake.wheel` on. No `wheel.platlib = False` builds.
+- Using cmake in SDist step is not supported yet.
+- Editable installs are not supported yet.
+- `scikit-build.generate` and `scikit-build.metadata` is not supported.
+- `${SKBUILD_HEADER_DIR}` is not supported, request support in Hatching if
+  needed.
+- Anything in `${SKBUILD_METADATA_DIR}` must be placed in an `extra_metadata`
+  folder.
+- Python 3.8 highly recommended as features are missing from the last hatchling
```
Comment (henryiii) [inline docs/plugins/hatchling.md:72]:
```suggestion
- Python 3.8 highly recommended as features are missing from the last Hatchling
```

File: docs/plugins/hatchling.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+  for 3.7.
```
Comment (henryiii) [inline docs/plugins/hatchling.md:73]:
```suggestion
  to support 3.7.
```

**Pull request merged**